### ISSUE TITLE
Why different workers (even on the same host) report significantly different global steps in the log?

### ISSUE THREAD

**Issue by yunjiangster** [horovod/horovod#2912]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 1.15.4
3. Horovod version: 0.21.3
4. MPI version: openmpi-4.1.1
5. CUDA version: 10.0
6. NCCL version: 2.13
7. Python version: 3.6.8
8. Spark / PySpark version: N/A
9. Ray version: N/A
10. OS and version: Centos 7.3
11. GCC version: 7.3
12. CMake version: 2.1.12

**Checklist:**
1. Did you search issues to find if somebody asked this question before? Yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)? NA
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)? NA
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?
Yes
**Bug report:**
Please describe erroneous behavior you're observing and steps to reproduce it.

My horovod runs fine with multi machine multi gpu on 6 machines each with 4 cards. The problem is I observed that the global steps on each worker can be drastically different, as shown below. My understanding is that Ring AllReduce is synchronous, so this should not have happened right? This raises concern about correctness of the underlying gradient update algorithm. 

> [1,19]<stderr>:I0518 01:45:51.842797 140673246885696 basic_session_run_hooks.py:260] loss = 0.63453865, step = 110300 (34.143 sec)
> [1,16]<stderr>:I0518 01:45:52.469465 140435215157056 basic_session_run_hooks.py:260] loss = 0.62881386, step = 117800 (32.090 sec)
> [1,14]<stderr>:I0517 09:46:12.817372 140134417708864 basic_session_run_hooks.py:260] loss = 0.6413416, step = 117500 (32.109 sec)
> [1,11]<stderr>:I0518 01:45:53.382837 140121191044928 basic_session_run_hooks.py:260] loss = 0.6546766, step = 84000 (45.546 sec)
> [1,9]<stderr>:I0518 01:45:54.767767 140267109783360 basic_session_run_hooks.py:260] loss = 0.6558395, step = 111600 (33.596 sec)
> [1,17]<stderr>:I0518 01:45:56.109468 140168733341504 basic_session_run_hooks.py:260] loss = 0.5943562, step = 116300 (32.527 sec)
> [1,22]<stderr>:I0518 01:45:58.320261 140665546368832 basic_session_run_hooks.py:260] loss = 0.6153215, step = 117800 (32.011 sec)
> [1,4]<stderr>:I0518 01:46:01.718719 139773529175872 basic_session_run_hooks.py:260] loss = 0.62689364, step = 117600 (31.897 sec)
> [1,6]<stderr>:I0518 01:46:06.128803 140386233333568 basic_session_run_hooks.py:260] loss = 0.6403286, step = 118200 (31.860 sec)
> [1,3]<stderr>:I0518 01:46:06.554964 140538349205312 basic_session_run_hooks.py:260] loss = 0.6023634, step = 117800 (32.052 sec)
> [1,2]<stderr>:I0518 01:46:07.081845 140416158574400 basic_session_run_hooks.py:260] loss = 0.63183314, step = 118400 (31.799 sec)
> [1,13]<stderr>:I0517 09:46:28.226042 140497101571904 basic_session_run_hooks.py:260] loss = 0.6281785, step = 116600 (32.351 sec)
> [1,15]<stderr>:I0517 09:46:29.313174 139719203542848 basic_session_run_hooks.py:260] loss = 0.6320455, step = 110400 (34.138 sec)
> [1,0]<stderr>:I0518 01:46:09.499719 140196336703296 basic_session_run_hooks.py:692] global_step/sec: 3.1392
> [1,0]<stderr>:I0518 01:46:09.501245 140196336703296 basic_session_run_hooks.py:260] loss = 0.6096021, step = 117600 (31.855 sec)
> [1,1]<stderr>:I0518 01:46:10.434172 140288825235264 basic_session_run_hooks.py:260] loss = 0.6586847, step = 118400 (31.840 sec)
> [1,18]<stderr>:I0518 01:46:12.555037 140537019397952 basic_session_run_hooks.py:260] loss = 0.59331954, step = 118600 (31.767 sec)
> [1,5]<stderr>:I0518 01:46:14.201884 139910391580480 basic_session_run_hooks.py:260] loss = 0.6119241, step = 117600 (31.947 sec)
> [1,12]<stderr>:I0517 09:46:35.603423 140213217134400 basic_session_run_hooks.py:260] loss = 0.63698995, step = 118800 (31.730 sec)
> [1,8]<stderr>:I0518 01:46:21.641412 140234710558528 basic_session_run_hooks.py:260] loss = 0.5902337, step = 118700 (31.761 sec)
> [1,20]<stderr>:I0518 01:46:22.082927 139952352003904 basic_session_run_hooks.py:260] loss = 0.6045385, step = 116900 (32.244 sec)
> [1,7]<stderr>:I0518 01:46:22.832490 139673803372352 basic_session_run_hooks.py:260] loss = 0.62267864, step = 76000 (49.848 sec)
> [1,21]<stderr>:I0518 01:46:23.584347 140209280739136 basic_session_run_hooks.py:260] loss = 0.6044893, step = 117800 (31.982 sec)
> [1,16]<stderr>:I0518 01:46:24.621129 140435215157056 basic_session_run_hooks.py:260] loss = 0.63424665, step = 117900 (32.152 sec)
> [1,14]<stderr>:I0517 09:46:44.909708 140134417708864 basic_session_run_hooks.py:260] loss = 0.60744435, step = 117600 (32.092 sec)
> [1,19]<stderr>:I0518 01:46:26.165857 140673246885696 basic_session_run_hooks.py:260] loss = 0.62028193, step = 110400 (34.323 sec)
> [1,17]<stderr>:I0518 01:46:28.486097 140168733341504 basic_session_run_hooks.py:260] loss = 0.63778627, step = 116400 (32.377 sec)
> [1,9]<stderr>:I0518 01:46:28.522030 140267109783360 basic_session_run_hooks.py:260] loss = 0.5903483, step = 111700 (33.754 sec)
> [1,22]<stderr>:I0518 01:46:30.288945 140665546368832 basic_session_run_hooks.py:260] loss = 0.6343375, step = 117900 (31.969 sec)
> [1,4]<stderr>:I0518 01:46:33.725464 139773529175872 basic_session_run_hooks.py:260] loss = 0.63947105, step = 117700 (32.007 sec)
> [1,23]<stderr>:I0518 01:46:33.893916 140579294136128 basic_session_run_hooks.py:260] loss = 0.6487763, step = 83000 (46.197 sec)
> [1,6]<stderr>:I0518 01:46:38.000238 140386233333568 basic_session_run_hooks.py:260] loss = 0.6085214, step = 118300 (31.871 sec)
> [1,11]<stderr>:I0518 01:46:38.438935 140121191044928 basic_session_run_hooks.py:260] loss = 0.61565, step = 84100 (45.056 sec)
> [1,3]<stderr>:I0518 01:46:38.665583 140538349205312 basic_session_run_hooks.py:260] loss = 0.6215742, step = 117900 (32.111 sec)
> [1,2]<stderr>:I0518 01:46:38.867290 140416158574400 basic_session_run_hooks.py:260] loss = 0.63392586, step = 118500 (31.785 sec)
> [1,13]<stderr>:I0517 09:47:00.538572 140497101571904 basic_session_run_hooks.py:260] loss = 0.60433674, step = 116700 (32.313 sec)
> [1,0]<stderr>:I0518 01:46:41.426178 140196336703296 basic_session_run_hooks.py:692] global_step/sec: 3.1322
> [1,0]<stderr>:I0518 01:46:41.427805 140196336703296 basic_session_run_hooks.py:260] loss = 0.6036617, step = 117700 (31.927 sec)
> [1,1]<stderr>:I0518 01:46:42.260947 140288825235264 basic_session_run_hooks.py:260] loss = 0.6456109, step = 118500 (31.827 sec)
> [1,15]<stderr>:I0517 09:47:03.691321 139719203542848 basic_session_run_hooks.py:260] loss = 0.6230123, step = 110500 (34.378 sec)
> [1,18]<stderr>:I0518 01:46:44.328350 140537019397952 basic_session_run_hooks.py:260] loss = 0.64912814, step = 118700 (31.773 sec)
> [1,5]<stderr>:I0518 01:46:46.188978 139910391580480 basic_session_run_hooks.py:260] loss = 0.60555613, step = 117700 (31.987 sec)
> [1,10]<stderr>:I0518 01:46:46.219042 140403209897792 basic_session_run_hooks.py:260] loss = 0.6482359, step = 36100 (107.608 sec)
> [1,12]<stderr>:I0517 09:47:07.323107 140213217134400 basic_session_run_hooks.py:260] loss = 0.65038604, step = 118900 (31.720 sec)
> [1,8]<stderr>:I0518 01:46:53.392455 140234710558528 basic_session_run_hooks.py:260] loss = 0.6199914, step = 118800 (31.751 sec)
> [1,20]<stderr>:I0518 01:46:54.295858 139952352003904 basic_session_run_hooks.py:260] loss = 0.62220246, step = 117000 (32.213 sec)
> [1,21]<stderr>:I0518 01:46:55.561463 140209280739136 basic_session_run_hooks.py:260] loss = 0.6194153, step = 117900 (31.977 sec)
> [1,16]<stderr>:I0518 01:46:56.752912 140435215157056 basic_session_run_hooks.py:260] loss = 0.61981446, step = 118000 (32.132 sec)
> [1,14]<stderr>:I0517 09:47:16.957984 140134417708864 basic_session_run_hooks.py:260] loss = 0.62638634, step = 117700 (32.048 sec)
> [1,19]<stderr>:I0518 01:47:00.282531 140673246885696 basic_session_run_hooks.py:260] loss = 0.61887693, step = 110500 (34.117 sec)

**Comment by yunjiangster** [comment]
Any comment or pointer? Could it be that horovod implement some sort of straggler skipping/optimization? I haven't dared to look into the internal Ring Allreduce implementation yet but would be happy to with some code pointer.
